src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java

```diff
@@ -48,29 +57,40 @@
                         TunnelTag.YES)
                 || Validators.isOfType(edge, AreaTag.class, AreaTag.YES)
                 || YES_VALUE.equals(edge.tag(INDOOR_KEY))
-                || Validators.isOfType(edge, HighwayTag.class, HighwayTag.SERVICE)
+                || highwayServiceValidator.test(edge)
                         && Validators.isOfType(edge, ServiceTag.class, ServiceTag.DRIVEWAY)
-                || edge.connectedNodes().stream().anyMatch(
-                        node -> Validators.isOfType(node, EntranceTag.class, EntranceTag.YES)
-                                || Validators.isOfType(node, AmenityTag.class,
-                                        AmenityTag.PARKING_ENTRANCE)));
+                || edge.connectedNodes().stream().anyMatch(node -> Validators.isOfType(node,
+                        EntranceTag.class, EntranceTag.YES)
+                        || Validators.isOfType(node, AmenityTag.class, AmenityTag.PARKING_ENTRANCE))
+                // Ignore edges with nodes containing Barrier tags
+                || edge.getAtlas()
+                        .nodesWithin(edge.bounds(),
+                                node -> Validators.hasValuesFor(node, BarrierTag.class))
+                        .iterator().hasNext());
     }
 
-    private static Predicate<Edge> intersectsCoreWayInvalidly(final Area building)
+    private Predicate<Edge> intersectsCoreWayInvalidly(final Area building)
     {
-        // An invalid intersection is determined by checking that the highway is a core way
-        return edge -> HighwayTag.isCoreWay(edge)
+        // An invalid intersection is determined by checking that its highway tag is in the
+        // TaggableFilter
+        return edge -> this.highwayFilter.test(edge)
                 // And if the edge intersects the building polygon
                 && edge.asPolyLine().intersects(building.asPolygon())
                 // And ignore intersections where edge has highway=service and building has
                 // Amenity=fuel
-                && !(Validators.isOfType(edge, HighwayTag.class, HighwayTag.SERVICE)
+                && !(highwayServiceValidator.test(edge)
                         && Validators.isOfType(building, AmenityTag.class, AmenityTag.FUEL))
+                // And ignore intersections where building has nodes within it with Amenity=fuel
+                && !edge.getAtlas().pointsWithin(building.asPolygon(),
```

Comment:
You could potentially wrap this  Iterator.streams() instead of .interator().hasNext. Might be more efficient with large Edges

Reply:
`pointsWithin()` return an Iterable of points that meet the specified condition. `iterator.hasNext()` is looking into that Iterable. Will converting this into stream have an impact even if the edge is too large as the points that meet this conditions would be ideally low.